README.md
# Piwigo-API

![](https://github.com/kkdad/piwigo-api/workflows/Ruby/badge.svg)  [![codecov](https://codecov.io/gh/KKDad/piwigo-api/branch/master/graph/badge.svg)](https://codecov.io/gh/KKDad/piwigo-api)  ![](https://github.com/kkdad/piwigo-api/workflows/Ruby%20Gem/badge.svg) [![Gem Version](https://badge.fury.io/rb/piwigo-api.svg)](https://badge.fury.io/rb/piwigo-api)  

[Piwigo](https://piwigo.org/) is open source web application to manage your collection of photos, and other medias. Piwigo provides an API for interacting with it. 

This is a ruby-based client for interacting with a Piwigo instance using the Piwigo API.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'piwigo-api'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install piwigo-api

## Usage

Just include 'piwigo/session' and the related classes, then querying Piwigo is fairly quick and straightforward. The latest RubyDocs for [piwigo-api](https://rubydoc.info/github/KKDad/piwigo-api) can be loacated on [rubydoc.info](https://rubydoc.info).

### Get the second album and all of it's children

```ruby
require 'piwigo/session'
require 'piwigo/albums'

session = Piwigo::Session.login('mypiwigo.fqdn', 'Adrian', 'mypassword', https: false) 
unless session.nil? 
    albums = Piwigo::Albums.list(session, album_id: 2, recursive: true)
    albums.each { |album| p "#{album.id}, #{album.name} - # of photos: #{album.total_nb_images}" }
end
```

### Add a new album
```ruby
require 'piwigo/session'
require 'piwigo/albums'

session = Piwigo::Session.login('10.100.230.78', 'Adrian', 'mypassword', https: false) 
unless session.nil? 
    album = Piwigo::Albums::Album.new
    album.name = "My First Album"
    album = Piwigo::Albums.add(session, album)
end
```

### List the Contents of an Album

```ruby
require 'piwigo/session'
require 'piwigo/images'

session = Piwigo::Session.login('mypiwigo.fqdn', 'Adrian', 'mypassword', https: false) 
unless session.nil? 
    results = Piwigo::Images.getImages(session, album_id: 4)
    results[:images].each { |image| p "#{image.id}, #{image.name} - #{image.element_url}" }
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/kkdad/piwigo-ruby.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

0.5.9 - 2021-07-13 
- Update version to fix github actions

0.5.8 - 2021-07-13 
- Update dependencies address CVE-2021-32740

0.5.7 - 2020-03-01
- Fixed open-ended dependency on codecov

0.5.6 - 2020-03-01
- Update dependencies to address security vulnerability

0.5.5 - 2019-11-26
- Fix another encoding issue when synchronizing folder with accented characters.

0.5.4 - 2019-11-26
- Fix crash in sniff_attributes when the image dosn't contain a valid EXIF data
- Fix parent albums when syncronizing a folder-tree of albums into Piwigo

0.5.3 - 2019-11-25
- Fixed to coverage reports
- Only generate coverage reports during CI build
- codecov.io seems to be ignoring some test files in the coverage report
- Use EXIF data to pull out the original image date when uploading to Piwigo

0.5.1 - 2019-11-23
- Fixed minor bug in image.lookup

0.5.0 - 2019-11-23
- Added a directory of example scripts
- Added a folder-album syncronization methods

0.4.0 - 2019-11-20
- Add methods to check for the existance of a specific image and to upload images

0.3.1 - 2019-11-20
- Add a codecov.yaml for coverage reports

0.3.0 - 2019-11-17
- Added a documentation URL pointing to rubydoc.info
- Added build badges
- Added ability to page though the images in an album

0.2.0 - 2019-11-17
- Refactored the Session to make it easier to use
- Support adding/deleting & listing albums

0.1.1 - 2019-11-16 
- Initial Version, supports login/logout against a piwigo server
- Fixing spec consistency issues


test/image_test.rb
require 'test_helper'

class ImageTest < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::Piwigo::VERSION
  end

  def test_image_list
    uri = MiniTest::Mock.new
    uri.expect(:nil?, false)
    uri.expect(:host, 'fakehost.fqdn')
    uri.expect(:port, '8754')
    uri.expect(:request_uri, '/ws.php')

    session = MiniTest::Mock.new
    session.expect(:uri, uri)
    session.expect(:uri, uri)
    session.expect(:uri, uri)
    session.expect(:uri, uri)

    session.expect(:id, '2')

    response = MiniTest::Mock.new
    response.expect(:code, '200')
    response.expect(:body, '{"stat":"ok","result":{"paging":{"page":2,"per_page":2,"count":2,"total_count":"29"},"images":[{"id":39,"width":640,"height":480,"hit":0,"file":"pulirmen1.jpg","name":"pulirmen1","comment":null,"date_creation":"2001-03-20 15:01:27","date_available":"2019-11-07 15:48:24","page_url":"http:\/\/10.100.230.78\/picture.php?\/39","element_url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-70af44c1.jpg","derivatives":{"square":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-70af44c1-sq.jpg","width":120,"height":120},"thumb":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-70af44c1-th.jpg","width":144,"height":108},"2small":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-70af44c1-2s.jpg","width":240,"height":180},"xsmall":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-70af44c1-xs.jpg","width":432,"height":324},"small":{"url":"http:\/\/10.100.230.78\/_data\/i\/upload\/2019\/11\/07\/20191107154824-70af44c1-sm.jpg","width":576,"height":432},"medium":{"url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-70af44c1.jpg","width":"640","height":"480"},"large":{"url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-70af44c1.jpg","width":"640","height":"480"},"xlarge":{"url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-70af44c1.jpg","width":"640","height":"480"},"xxlarge":{"url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-70af44c1.jpg","width":"640","height":"480"}},"categories":[{"id":4,"url":"http:\/\/10.100.230.78\/index.php?\/category\/4","page_url":"http:\/\/10.100.230.78\/picture.php?\/39\/category\/4"}]},{"id":40,"width":1280,"height":960,"hit":0,"file":"quebec_city_in_snow_3.jpg","name":"quebec city in snow 3","comment":null,"date_creation":"2002-03-10 06:23:15","date_available":"2019-11-07 15:48:24","page_url":"http:\/\/10.100.230.78\/picture.php?\/40","element_url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-e45605e1.jpg","derivatives":{"square":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-sq.jpg","width":120,"height":120},"thumb":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-th.jpg","width":144,"height":108},"2small":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-2s.jpg","width":240,"height":180},"xsmall":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-xs.jpg","width":432,"height":324},"small":{"url":"http:\/\/10.100.230.78\/_data\/i\/upload\/2019\/11\/07\/20191107154824-e45605e1-sm.jpg","width":576,"height":432},"medium":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-me.jpg","width":792,"height":594},"large":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-la.jpg","width":1008,"height":756},"xlarge":{"url":"http:\/\/10.100.230.78\/i.php?\/upload\/2019\/11\/07\/20191107154824-e45605e1-xl.jpg","width":1224,"height":918},"xxlarge":{"url":"http:\/\/10.100.230.78\/upload\/2019\/11\/07\/20191107154824-e45605e1.jpg","width":"1280","height":"960"}},"categories":[{"id":4,"url":"http:\/\/10.100.230.78\/index.php?\/category\/4","page_url":"http:\/\/10.100.230.78\/picture.php?\/40\/category\/4"}]}]}}')

    httpclient = MiniTest::Mock.new
    httpclient.expect(:request, response, [Net::HTTP::Post])

    Net::HTTP.stub(:new, httpclient) do
      results = Piwigo::Images.getImages(session, album_id: 4)

      refute results.nil?
      images = results[:images]
      refute images.nil?
      assert images.size == 2
      assert images[0].id == 39
      assert images[0].name == 'pulirmen1'
      paging = results[:paging]
      refute paging.nil?
    end

    uri.verify
    session.verify
    response.verify
    httpclient.verify
  end

  def test_lookup
    uri = MiniTest::Mock.new
    uri.expect(:nil?, false)
    uri.expect(:host, 'fakehost.fqdn')
    uri.expect(:port, '8754')
    uri.expect(:request_uri, '/ws.php')

    session = MiniTest::Mock.new
    session.expect(:uri, uri)
    session.expect(:uri, uri)
    session.expect(:uri, uri)
    session.expect(:uri, uri)

    session.expect(:id, '2')

    response = MiniTest::Mock.new
    response.expect(:code, '200')
    response.expect(:body, '{"stat":"ok","result":{"c8e8758dbfab0f0fa14c44edee1da8ad":"319"}}')

    httpclient = MiniTest::Mock.new
    httpclient.expect(:request, response, [Net::HTTP::Post])

    file = 'fakeimage.png'
    File.stub(:binread, 'fakeimage file contents') do
      Net::HTTP.stub(:new, httpclient) do
        results = Piwigo::Images.lookup(session, file)       

        refute results.nil?
        assert results == '319'
      end
    end

    uri.verify
    session.verify
    response.verify
    httpclient.verify
  end

  def test_lookup_not_present
    uri = MiniTest::Mock.new
    uri.expect(:nil?, false)
    uri.expect(:host, 'fakehost.fqdn')
    uri.expect(:port, '8754')
    uri.expect(:request_uri, '/ws.php')

    session = MiniTest::Mock.new
    session.expect(:uri, uri)
    session.expect(:uri, uri)
    session.expect(:uri, uri)
    session.expect(:uri, uri)

    session.expect(:id, '2')

    response = MiniTest::Mock.new
    response.expect(:code, '200')
    response.expect(:body, '{"stat":"ok","result":{"c8e8758dbfab0f0fa14c44edee1da8ad":null}}')

    httpclient = MiniTest::Mock.new
    httpclient.expect(:request, response, [Net::HTTP::Post])

    file = 'fakeimage.png'
    File.stub(:binread, 'fakeimage file contents') do
      Net::HTTP.stub(:new, httpclient) do
        results = Piwigo::Images.lookup(session, file)

        assert results.nil?
      end
    end

    uri.verify
    session.verify
    response.verify
    httpclient.verify
  end


end


test/piwigo_test.rb
require "test_helper"

class PiwigoTest < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::Piwigo::VERSION
  end

  def test_login_logoff_success
    logon = MiniTest::Mock.new
    logon.expect(:code, '200')
    logon.expect(:body, '{"stat":"ok","result":true}')
    logon.expect(:response, {'set-cookie' => 'pwg_id=ns4bilhgbr2nrhk0to226fs2cc; path=/; HttpOnly, pwg_remember=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/'})

    logout = MiniTest::Mock.new
    logout.expect(:code, '200')
    logout.expect(:body, '{"stat":"ok","result":true}')

    httpclient = MiniTest::Mock.new
    httpclient.expect(:request, logon, [Net::HTTP::Post])
    httpclient.expect(:request, logout, [Net::HTTP::Get])

      Net::HTTP.stub(:new, httpclient) do
        session = Piwigo::Session.new
        session.login('mygallery.fakefqdn.com', 'myuser', 'mypassword', https: false)

        assert session.session_id == ["pwg_id=ns4bilhgbr2nrhk0to226fs2cc"]

        session.logout
        assert session.session_id.nil?
      end
      logon.verify
      logout.verify
      httpclient.verify
  end


end


lib/piwigo/albums.rb
require 'net/http'
require 'uri'
require 'json'
require 'logger'

# Piwigo organizes images by albums. The album tree has unlimted depth and each photo can belong to multiple albums. The Piwigo API
# refers to a Album as a Category.
module Piwigo
  class Albums
    class Album
      # @return [Number] Album ID
      attr_accessor :id

      # @return [String] Name of the Album
      attr_accessor :name

      # @return [String] Album Description
      attr_accessor :comment

      # ???
      attr_accessor :permalink

      # @return [String] public or private
      attr_accessor :status

      # ???
      attr_accessor :uppercats

      # The rank of an album
      attr_accessor :global_rank

      # @return [Number] ID of the parent album, nil of this is a top-level album (Aka ParentID)
      attr_accessor :id_uppercat

      # @return [Number] Number of the Newest photos, excluding child albums
      attr_accessor :nb_images

      # @return [Number] Number of the Newest photos, including child albums
      attr_accessor :total_nb_images

      # ID of the representative photo for an album. This photo doesn't have to belong to the album.
      attr_accessor :representative_picture_id

      # @return [DateTime] Date of the Newest photos, excluding child albums
      attr_accessor :date_last

      # @return [DateTime] Date of the Newest photos, including child albums
      attr_accessor :max_date_last

      # @return [Number] Number of child albums inside this album
      attr_accessor :nb_categories

      # @return [Atring] Album URL
      attr_accessor :url

      # @return [String] Thumbnail URL
      attr_accessor :tn_url

      def initialize(hash: nil)
        hash&.each do |key, value|
          # Bug: If the encoding is Windows-1252, then Piwigo will blowup when creating the album
          value = value.encode('UTF-8', 'Windows-1252') if value.class == String && value.encoding.to_s == 'Windows-1252'
          send("#{key}=", value)
        end
      end

      def to_s
        "#{name}(#{id})"
      end
    end

    # Returns a list of albums
    #
    # @param [Session] session - Session
    # @param [Number] album_id - Album to fetch, Optional
    # @param [Boolean] recursive - Include subalbums, Optional
    # @param [Boolean] public - Only include public albums, Optional
    # @param [Boolean] fullname - ???, Optional
    # @param [String] thumbnail_size - Size of thumbname to return, One of: square, thumb, 2small, xsmall, small, medium, large, xlarge, xxlarge. Optional
    # @param [Logger] logger logger to output debug messages to (Optional)
    #
    # @return [Array<Album>] All albums that match the criteria, or nil there were no matches
    def self.list(session, album_id: nil, recursive: nil, public: nil, fullname: nil, thumbnail_size: nil, logger: nil)
      raise 'Invalid session' if session.uri.nil?

      logger ||= Logger.new(STDOUT)

      begin
        http = Net::HTTP.new(session.uri.host, session.uri.port)
        request = Net::HTTP::Post.new(session.uri.request_uri)
        form = {
          method: 'pwg.categories.getList'
        }
        form[:cat_id] = album_id unless album_id.nil?
        form[:recursive] = recursive unless recursive.nil?
        form[:public] = public unless public.nil?
        form[:fullname] = fullname unless fullname.nil?
        form[:thumbnail_size] = thumbnail_size unless thumbnail_size.nil?
        request.set_form_data(form)
        request['Cookie'] = [session.id]

        # Send the request
        response = http.request(request)
        if response.code == '200'
          data = JSON.parse(response.body)
          albums = data['result']['categories'].map { |hash| Album.new(hash: hash) }
          logger.info "Album List succeeded: #{albums.size} albums retrieved."
          albums
        end
      rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
        logger.error "Album List failed: #{e.messages}"
        nil
      end
    end

    # Lookup a specific Album from a list of albums
    #
    # @param [<Type>] session
    # @param [<Type>] album_name - Name of the album to locate
    # @param [<Type>] logger <description>
    #
    # @return [<Type>] Album if located, nil otherwise
    def self.lookup(session, album_name, logger: nil)
      albums = list(session, recursive: true, fullname: false, logger: logger)
      filtered = albums.select do |album|
        album.name == album_name
      end
      filtered.size == 1 ? filtered[0] : nil
    end

    # Adds an album.
    #
    # @param [Session] session to interact with Piwigo
    # @param [Album] album album to create
    # @param [Logger] logger logger to output debug messages to (Optional)
    #
    # @return [Album] newly created album
    def self.add(session, album, logger: nil)
      raise 'Invalid session' if session.uri.nil?
      raise 'Invalid album' if album.nil?

      logger ||= Logger.new(STDOUT)

      begin
        http = Net::HTTP.new(session.uri.host, session.uri.port)
        request = Net::HTTP::Post.new(session.uri.request_uri)
        logger.info "Encoding: #{album.name} - #{album.name.encoding}"
        form = {
          method: 'pwg.categories.add',
          name: album.name
        }
        form[:parent] = album.id_uppercat unless album.id_uppercat.nil?
        form[:comment] = album.comment unless album.comment.nil?
        form[:status] = album.status unless album.status.nil?
        request.set_form_data(form)
        request['Cookie'] = [session.id]

        # Send the request
        response = http.request(request)
        if response.code == '200'
          data = JSON.parse(response.body)
          album.id = data['result']['id']
          logger.info "Album Add succeeded: #{album.name}(#{album.id}) created."
          album
        end
      rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
        logger.error "Album Add failed: #{e.messages}"
        nil
      end
    end

    # Deletes album(s).
    #
    # @param [Session] piwigo session
    # @param [Number] id of the album to remove
    # @param [String] photo_deletion_mode can be "no_delete" (may create orphan photos), "delete_orphans" (default mode, only deletes photos linked to no other album)
    #                 or "force_delete" (delete all photos, even those linked to other albums)
    # @param [Logger] logger logger to output debug messages to (Optional)
    #
    # @return [Boolean] true if album was deleted
    def self.delete(session, id, photo_deletion_mode: nil, logger: nil)
      raise 'Invalid session' if session.uri.nil?

      logger ||= Logger.new(STDOUT)

      begin
        http = Net::HTTP.new(session.uri.host, session.uri.port)
        request = Net::HTTP::Post.new(session.uri.request_uri)
        request.body = "method=pwg.categories.delete&category_id=#{id}"
        request.body.concat "&photo_deletion_mode=#{photo_deletion_mode}" unless photo_deletion_mode.nil?
        request.body.concat "&pwg_token=#{session.pwg_token}"
        request['Cookie'] = [session.id]

        # Send the request
        response = http.request(request)
        if response.code == '200'
          data = JSON.parse(response.body)
          logger.info "Album Delete succeeded: Album #{id} removed - #{data}"
          true
        else
          p response.code
          p response.body
          false

        end
      rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
        logger.error "Album delete: #{e.messages}"
        false
      end
    end
  end
end


lib/piwigo/images.rb
require 'base64'
require 'net/http'
require 'uri'
require 'json'
require 'logger'
require 'digest'
require_relative 'image_uploader'

# Piwigo organizes images by albums. The album tree has unlimted depth and each photo can belong to multiple albums. The Piwigo API
# refers to a Album as a Category.
module Piwigo
  class Images
    class Image
      # @return [Number] Unique ID idenifying this ie
      attr_accessor :id

      # @return [Number] Width of the image in pixels
      attr_accessor :width

      # @return [Number] Height of the image in pixels
      attr_accessor :height

      # @return [Number] Number of times the image has been viewed
      attr_accessor :hit

      # @return [String] Filename for the image
      attr_accessor :file

      # @return [String] Name of the image
      attr_accessor :name

      # @return [String] Comments about the image
      attr_accessor :comment

      # @return [DateTime] DateTime when the image was taken
      attr_accessor :date_creation

      # @return [DateTime] DateTime when the image was uploaded to Piwigo
      attr_accessor :date_available

      # @return [String] URL to the image page
      attr_accessor :page_url

      # @return [String] URL to the image itself
      attr_accessor :element_url

      # @return [Array<String>] Links to different sizes of the image
      attr_accessor :derivatives

      # @return [<Type>] List of all of the Albums this image is in
      attr_accessor :categories

      def initialize(hash: nil)
        hash&.each do |key, value|
          # Bug: If the encoding is Windows-1252, then Piwigo will blowup when creating the album
          value = value.encode('UTF-8', 'Windows-1252') if value.class == String && value.encoding.to_s == 'Windows-1252'
          send("#{key}=", value)
        end
      end
    end

    class Paging
      # @return [Number] Page number of the results
      attr_accessor :page

      # @return [Number] Number of images requesed per page
      attr_accessor :per_page

      # @return [Number] Number of Images on this page. When this is less then the number per page, then there are no more results.
      attr_accessor :count

      # @return [Number] Total number of images across all pages
      attr_accessor :total_count

      def initialize(hash: nil)
        hash&.each do |key, value|
          # Bug: If the encoding is Windows-1252, then Piwigo will blowup when creating the album
          value = value.encode('UTF-8', 'Windows-1252') if value.class == String && value.encoding.to_s == 'Windows-1252'
          send("#{key}=", value)
        end
      end
    end

    # Returns elements for the corresponding categories.
    # order comma separated fields for sorting
    #
    # @param [Session] session
    # @param [Number] album_id - Can be empty if recursive is true.
    # @param [Boolean] recursive - Include images from child albums
    # @param [Number] per_page - Number of items to include per page
    # @param [Number] page - Page to retrieve
    # @param [String] order - One or more of id, file, name, hit, rating_score, date_creation, date_available or random
    # @param [Logger] logger logger to output debug messages to (Optional)
    #
    # @return [Hash] <description>
    def self.getImages(session, album_id: nil, recursive: nil, per_page: 100, page: 0, order: nil, logger: nil)
      raise 'Invalid session' if session.uri.nil?

      logger ||= Logger.new(STDOUT)

      begin
        http = Net::HTTP.new(session.uri.host, session.uri.port)
        request = Net::HTTP::Post.new(session.uri.request_uri)
        form = {
          method: 'pwg.categories.getImages'
        }
        form[:cat_id] = album_id unless album_id.nil?
        form[:recursive] = recursive unless recursive.nil?
        form[:per_page] = per_page unless per_page.nil?
        form[:order] = order unless order.nil?
        form[:page] = page unless page.nil?
        request.set_form_data(form)
        request['Cookie'] = [session.id]

        # Send the request
        response = http.request(request)
        if response.code == '200'
          data = JSON.parse(response.body)
          paging = Paging.new hash: data['result']['paging']
          images = data['result']['images'].map { |hash| Image.new(hash: hash) }
          logger.info "Image List succeeded: #{images.size} images retrieved."
          { paging: paging, images: images }
        end
      rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
        logger.error "Image List failed: #{e.message}"
        nil
      end
    end

    # Add a photo to Piwigo
    #
    # @param [<Type>] session
    # @param [<Type>] filename of the file to upload
    # @param [<Type>] name of the image
    #
    # @return [Boolean] True if successful
    def self.upload(session, file, name)
      ImageUploader.new(session, file, name).upload
    end

    # Checks existence of images
    #
    # @param [Session] session
    # @param [String] file to check
    # @param [Logger] logger
    #
    # @return [Number] Piwigo image_id if matched, nil if not present
    def self.lookup(session, file, logger: nil)
      raise 'Invalid session' if session.uri.nil?

      logger ||= Logger.new(STDOUT)
      image_content = File.binread(file)
      file_sum = Digest::MD5.hexdigest(image_content)

      begin
        http = Net::HTTP.new(session.uri.host, session.uri.port)
        request = Net::HTTP::Post.new(session.uri.request_uri)
        form = {
          method: 'pwg.images.exist',
          md5sum_list: file_sum,
          filename_list: file
        }
        request.set_form_data(form)
        request['Cookie'] = [session.id]

        # Send the request
        response = http.request(request)
        if response.code == '200'
          data = JSON.parse(response.body)
          logger.info "Image lookup succeeded: #{data['result']}"
          data['result'][file_sum]
        end
      rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
        logger.error "Image lookup failed: #{e.message}"
        nil
      end
    end
  end
end


lib/piwigo/folder_sync.rb
require 'logger'
require 'find'
require 'exifr/jpeg'
require_relative 'albums'
require_relative 'image_uploader'

module Piwigo
  # Syncronize a folder with Piwigo
  #    - Album name will be the same as the folder name
  #    - Will be created as a top-level folder unless specified
  class FolderSync
    # Create a new ImageUploader
    #
    # @param [Session] session - session to a piwigo instance
    def initialize(session, logger: nil)
      @logger = logger || Logger.new(STDOUT)
      @session = session
      @current_album = nil
      @parent_album = nil
    end

    def self.synchronize(session, directory, logger: nil)
      unless File.directory?(directory)
        @Logger.error "No such directory: #{directory}"
        return false
      end
      FolderSync.new(session, logger: logger).synchronize(directory)
    end

    # Syncronize a folder with Piwigo
    #
    # @param [String] directory - directory to  syncronize.
    def synchronize(directory)
      if ['originals', '.picasaoriginals'].include? File.basename(directory.downcase)
        @logger.info "Skipping special directory: #{directory}"
        return
      else
        @logger.info "Processing Directory: #{directory}"
      end

      Dir.entries(directory).reject { |entry| entry =~ /^.{1,2}$/ }.each do |directory_entry|
        item_to_process = (File.join directory, directory_entry).encode('utf-8')

        if File.directory? item_to_process
          @parent_album = ensure_album(File.dirname(item_to_process), nil)
          @current_album = ensure_album(item_to_process, @parent_album)

          # Recursively process all of the entries in this album
          synchronize item_to_process
        else
          process_file item_to_process
        end
      end
    end

    private

    def ensure_album(directory_entry, parent_album)
      album_name = File.basename(directory_entry)
      @logger.info "Ensuring Album #{album_name} exits"
      return if ['/', '\\', ''].include? album_name

      # Look to see if we have previously created an Album in Piwigo with this name
      album = Piwigo::Albums.lookup(@session, album_name)

      if album.nil?
        # Album doesn't exist, create one
        new_album = { 'name' => album_name,
                      'id_uppercat' => parent_album.nil? ? nil : parent_album.id }
        album = Piwigo::Albums::Album.new(hash: new_album)
        album = Piwigo::Albums.add @session, album
      end
      album
    end

    def process_file(directory_entry)
      # Only attempt to import images
      return unless ['.jpg', '.png', '.gif'].include? File.extname(directory_entry).downcase

      @logger.info "Processing Image: '#{directory_entry}' in album '#{@current_album}'"
      image = Piwigo::Images.lookup(@session, directory_entry)
      return unless image.nil?

      Piwigo::Images::ImageUploader.new(logger: @logger).upload(@session, directory_entry, File.basename(directory_entry), album: @current_album)
    end
  end
end


lib/piwigo/session.rb
require 'net/http'
require 'uri'
require 'json'
require 'logger'

module Piwigo
    # Class to hold the Piwigo session the rest of the API needs to pass in order to access the API
    class Session
        
        # @return [String] Piwigo session cookie
        attr_accessor :id

        # @return [URI::HTTP] Piwigo host associated with this session
        attr_accessor :uri

        # @return [String] token required for admin methods
        attr_accessor :pwg_token

        # @return [String] token required for admin methods
        attr_accessor :username


        def initialize(id, uri)
            self.id = id
            self.uri = uri

            status            
        end
          
        # Gets information about the current session. Also provides a token useable with admin methods.
        # {"stat":"ok","result":{"username":"Adrian","status":"webmaster","theme":"modus","language":"en_GB","pwg_token":"9edde6a1ae535934cca6a2423f9bcbe7","charset":"utf-8","current_datetime":"2019-11-17 21:57:58","version":"2.10.1","available_sizes":["square","thumb","2small","xsmall","small","medium","large","xlarge","xxlarge"],"upload_file_types":"jpg,jpeg,png,gif","upload_form_chunk_size":500}}
        # @return [<Type>] <description>
        def status
            logger = logger || Logger.new(STDOUT)     

            begin                       
                # Create the HTTP objects
                http = Net::HTTP.new(uri.host, uri.port)
                request = Net::HTTP::Post.new(uri.request_uri)
                request.body = "method=pwg.session.getStatus"
                request['Cookie'] = self.id
                
                # Send the request
                response = http.request(request)    
            
                if response.code == '200'
                    logger.info "Status succeeded"
                    data = JSON.parse(response.body)
                    self.pwg_token = data['result']['pwg_token']
                    self.username = data['result']['username']
                end

            rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
                    logger.error "Unexpected failed: #{e.messages}"
            end
            nil              
            
        end
           
        # Logout of the current session
        # @param [Logger] logger logger to output debug messages to (Optional)        
        def logout(logger: nil)            
            raise "This session has already been logged out" if uri.nil?
            logger = logger || Logger.new(STDOUT)     

            # Create the HTTP objects
            http = Net::HTTP.new(self.uri.host, self.uri.port)
            request = Net::HTTP::Get.new(self.uri.request_uri + "&method=pwg.session.logout")
            request['Cookie'] = self.id
            
            # Send the request
            response = http.request(request)               
            logger.info "Logout succeeded: #{response.body}" if response.code == '200'    
            self.id = nil
            self.uri = nil
        end
        
        
        # Log into the piwigo API and grab the session id for subsequent calls.
        # @param [string] piwigo - host to connect to. Can be fqdn or ip. 
        # @param [string] username - user to connect as
        # @param [string] password - password for user
        # @param [boolean] https - Use HTTPS?
        # @param [Logger] logger logger to output debug messages to (Optional)        
        def self.login(host, username, password, https: true, logger: nil)

            raise "host should not be nil" if host.nil?
            raise "username should not be nil" if username.nil?
            logger = logger || Logger.new(STDOUT)     

            begin
            
                uri = https ? URI::HTTPS.build(host: host, path: "/ws.php", query: "format=json") :
                                   URI::HTTP.build(host: host, path: "/ws.php", query: "format=json")                 
            
                # Create the HTTP objects
                http = Net::HTTP.new(uri.host, uri.port)
                request = Net::HTTP::Post.new(uri.request_uri)
                request.body = "method=pwg.session.login&username=#{username}&password=#{password}"
                
                # Send the request
                response = http.request(request)    
            
                if response.code == '200'
                    logger.info "Login succeeded: #{response.body}"
                    pwg_id = response.response['set-cookie'].split(';').select { |i| i.strip.start_with? "pwg_id" }.first
                    return Session.new(pwg_id, uri)
                else
                    logger.error "Login failed: #{response.body}"
                end

            rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
                    logger.error "Login failed: #{e.messages}"
            end
            nil                
        end       
    end
end